### PR TITLE
improve type support & rework templates:

### PR DESCRIPTION
**Pull request by iwanbk**
- add partial map support, using `additionalProperties` field
- fix array and enum support
- use Go 1.6 trimming spaces in template
- union implemented as `interface{}`
- type specialization fix 590554575d1bc74cbd7faf68d7785ee366bae6c6
- syntactic sugar & shortcut support

### CODE REVIEW

**Comment by iwanbk** [comment]
map supported using `additionalProperties` field.

Note that Map, Array, and Enum here are about Types not properties type.

Map in type : 
`type NewType map[string]string`

Map in properties: 
```
type SomeType struct {
   SomeField map[string]string
}

**Comment by iwanbk** [comment]
I have added limitation about map type in limitations.md. This PR is ready

**Comment by mohabusama** [comment]
Hi @iwanbk 

I was trying to test the map types generation on this branch:
assuming the following input
```YAML
#%RAML 1.0
title: UserApi
version: 0.1
baseUri: /users
types:
 emails:
    type: object
    additionalProperties: string
 user:
    properties:
        username:
            type: string
        emails:
            type: emails

    example:
        username: bob
        emails:
            work: bob@company.com
            home: bob@example.com
/users:
  post:
    description: Create a new user
    body:
        application/json:
            type: user
```
the generated `Emails` type is:

```go
package main

type Emails struct {
}
```
and the User is:

```go
package main

type User struct {
        Emails   emails `json:"emails"`
        Username string `json:"username"`
}
```
Did i miss something here?

Thanks
Mohab

**Comment by iwanbk** [comment]
@mohabusama 
Looks like output from old binary.

Have you rebuild the project? using `build.sh` script

**Comment by mohabusama** [comment]
Correct.

I was also using Go1.5, so I had errors with the template. I suggest adding sanity check in `build.sh`.

With the new binary `Emails` type was generated properly, but User was still the same:

```go
package main

type User struct {
        Emails   emails `json:"emails"`
        Username string `json:"username"`
}
```
I am expecting the User to look like this:

```go
package main

type User struct {
        Emails   Emails `json:"emails"`
        Username string `json:"username"`
}
```
as it is referencing the `Emails` type.

What do you think?

Thanks,
Mohab

File: commands/templates/server_resources_api.tmpl [hunk 1]
```diff
@@ -13,28 +13,27 @@ type {{.Name}}API struct {
...
 
-	{{ if .RespBody }}
+	{{- if .RespBody }}
 	var respBody {{.RespBody}}
 	json.NewEncoder(w).Encode(&respBody);
-	{{ end }}
-
+	{{- end }}
 	// uncomment below line to add header
 	// w.Header.Set("key","value")
```
Comment (mohabusama) [inline commands/templates/server_resources_api.tmpl]: This should be `w.Header().Set("key","value")`

**Comment by iwanbk** [comment]
```
as it is referencing the Emails type.

What do you think?
```
It is a bug.

**Comment by iwanbk** [comment]
@mohabusama 
```
as it is referencing the Emails type.
```
fixed in b4719c1cfc08e72d4100dbc16c6f47efa92d4dcd

File: commands/templates/server_resources_api.tmpl (hunk 1 above)
Comment (iwanbk) [inline commands/templates/server_resources_api.tmpl]: fixed in 3cffd8b964300828dc748ec3c961b9ac1e2b2f8b

**Comment by iwanbk** [comment]
union support added in 51a2dedca125d62bb22649ecea23dab98af178ad

**Comment by iwanbk** [comment]
type specialization support fixed at 590554575d1bc74cbd7faf68d7785ee366bae6c6.

@robvanmieghem @mohabusama this PR is ready again :)

**Comment by iwanbk** [comment]
@robvanmieghem 
syntactic sugar & shortcut support added in 5147ae9dd1c989756cd8b9b7f080d1e0eae0e744

**Pull request merged**